rfcs/20230418-constant-tensor-cache-capacity-control/README.md

```diff
@@ -0,0 +1,226 @@
+to query a constant tensor from the cache. If cache miss, the backend will add a
+promised future value to the cache and fulfill the promise after the tensor
+computation completes (The mentioned promise-future mechanism). The
+`remove_if_exist` method could be used when backend want to delete an item from
+the cache.
+
+When the cached tensor size reach the capacity, the new coming tensor won't be
+cached anymore by default. In later section, we will further discuss the policy
+about how to manage the cache size.
+
+The instantiated constant tensor cache instance will have global lifecycle and
+will be destroyed when the library destroyed.
+
+### Capacity setting/getting frontend API
+Based on above considerations, we proposed to add the following two C++ API and
+corresponding C API to control the capacity of constant tensor cache:
+
+```cpp
+void set_constant_tensor_cache_capacity(dnnl::engine::kind eng_kind, int size);
+int get_constant_tensor_cache_capacity(dnnl::engine::kind eng_kind);
+```
+
+The unit of size is `MBytes`. Users can call the setter API to set the capacity
+of constant tensor cache for specific engine kind. Setting capacity to 0 means
+clear all cached tensors. The getter API can be used to query the current
+capacity of specific engine kind. These two APIs can be called multiple times
+during the model running, but users should be aware that setting a smaller
+capacity may evict cached tensor and impact performance.The default capacity is
+`std::numeric_limits<size_t>::max()`, which means we can cache constant tensors
+with no capacity limitation.
+
+We also allow to control the capacity by using an environment variable named
+`ONEDNN_CONSTANT_TENSOR_CACHE_CAPACITY`. The form of this env var value is
+`engine_kind:size`. We can use `;` to split the configuration for different
+engine kind. Below is an example:
+
+```shell
+# set cpu/gpu constant tensor cache capacity to 10 GB and 2 GB separately
+export ONEDNN_CONSTANT_TENSOR_CACHE_CAPACITY=cpu:10240;gpu:2048 
+```
+
+### Cache management policy control
+As mentioned above, when the cache reaches capacity, the new coming tensor won't
+be cached anymore by default. We can called this policy as `NO_EVICT` simply.
+However, this policy may be not optimal sometimes. For example, generating
+different constant tensor may have different cost (like consumed time, memory,
+...), users may prioritize evicting those low cost constant tensors.
+
+Based on above considerations, we could extend the above capacity control API to
+allowing to set/get cache management policy. Then API users could try different
+policies for their own scenarios and select the best one.
+
+```cpp
+enum policy_t {
+    NO_EVICT,
+    COST,
+    ... // Others
+}
+
+void set_constant_tensor_cache_capacity(dnnl::engine::kind eng_kind, int size, policy_t policy);
+void get_constant_tensor_cache_capacity(dnnl::engine::kind eng_kind, int* size, policy_t * policy);
```

Comment:
Is there a way for backend to query the usage of cache capacity? Then backend can know the tensor cache is full, and we can generate new kernels in plain format to avoid using more tensor cache.

Reply:
I can try to add such an backend API if it's needed.

Reply:
added a `get_size` api for backend